### PR TITLE
recursive sparrow builder from nested "ArrayOfStruct" - like structures

### PR DESCRIPTION
**Pull request by DerThorsten**

### CODE REVIEW

File: include/sparrow/layout/array_base.hpp [hunk 1]
```diff
@@ -123,7 +123,7 @@ namespace sparrow
         array_crtp_base(const array_crtp_base&) = default;
         array_crtp_base& operator=(const array_crtp_base&) = default;
 
-        array_crtp_base(array_crtp_base&&) = default;
+        array_crtp_base(array_crtp_base&& other) = default;
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/layout/array_base.hpp:126]: leftoever from debugging...

File: include/sparrow/layout/list_layout/list_array.hpp [hunk 2]
```diff
@@ -509,7 +562,6 @@ namespace sparrow
         VB && validity_input
     )
     {
-        SPARROW_ASSERT(list_offsets.size() == list_sizes.size() , "sizes and offset must have the same size");
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/layout/list_layout/list_array.hpp:512]: probably an accident while debugging some issues

File: include/sparrow/builder/builder_utils.hpp
Comment (Alex-PLACET) [inline include/sparrow/builder/builder_utils.hpp:1]: Can you add tests for functions and concepts please?

Comment (DerThorsten) [reply include/sparrow/builder/builder_utils.hpp:1]: done

File: include/sparrow/builder/nested_eq.hpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                // a is null, b is not
+                else if (!a.has_value())
+                {
+                    return false;
+                }
+                // a is not null, b is
+                else if (!b.has_value())
+                {
+                    return false;
+                }
```
Comment (Alex-PLACET) [inline include/sparrow/builder/nested_eq.hpp:75]:
```suggestion
                // a or b is null
                else if (!a.has_value() || !b.has_value())
                {
                    return false;
                }
```

File: include/sparrow/layout/array_wrapper.hpp [hunk 4]
```diff
@@ -34,8 +34,8 @@ namespace sparrow
         t.end();
         t.cbegin();
         t.cend();
-        t.bitmap();
-        t.values();
+        //t.bitmap(); // NOPE
```
Comment (Alex-PLACET) [inline include/sparrow/layout/array_wrapper.hpp:37]: To remove ?

File: test/test_builder_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+                // the return type of for_each_index indicates if the loop was run to completion
+                detail::for_each_index<0>([](auto i) 
+                {
+                    // should not be called
+                    CHECK_FALSE(decltype(i)::value, 0);
+                    CHECK(false);
+                    return true; // should not be called
+                });
+            }
+            SUBCASE("non-empt")
```
Comment (Alex-PLACET) [inline test/test_builder_utils.cpp:55]:
```suggestion
            SUBCASE("non-empty")
```

File: include/sparrow/builder/builder.hpp [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+{
+    using type = sparrow::run_end_encoded_array;
+    using raw_range_value_type = std::ranges::range_value_t<T>;
+
+    template<class U>
+    static type create(U && t)
+    {   
+        using value_type = std::decay_t<raw_range_value_type>;
+
+        std::vector<value_type> values{};
```
Comment (Alex-PLACET) [inline include/sparrow/builder/builder.hpp:363]: CAn you reserve ?

File: include/sparrow/builder/builder.hpp [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+    // keep the nulls
+    using raw_range_value_type = std::ranges::range_value_t<T>;
+
+    template<class U>
+    static type create(U && t)
+    {   
+        key_type key = 0;
+        std::map<raw_range_value_type, key_type, nested_less<raw_range_value_type>> value_map;
+        std::vector<raw_range_value_type> values;
+        std::vector<key_type> keys;
```
Comment (Alex-PLACET) [inline include/sparrow/builder/builder.hpp:320]: Can you reserve (maybe at the possible maximum), then shrink

**Review by Alex-PLACET — APPROVED** [review]
Huge work, happy that you succeeded to made this feature available for all compilers

File: include/sparrow/builder/builder.hpp (hunk 7 above)
Comment (DerThorsten) [reply include/sparrow/builder/builder.hpp:320]: sure, but I think we can skip the shrink

File: include/sparrow/builder/builder.hpp
[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/builder/builder.hpp:1]: done

File: include/sparrow/builder/nested_eq.hpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            bool operator()(const T& a, const T& b) const
+            {
+                const auto size = range_size(a);
+                if(size != range_size(b))
+                {
+                    return false;
+                }
+                auto a_it = std::ranges::begin(a);
+                auto b_it = std::ranges::begin(b);
+                for(std::size_t i = 0; i < size; ++i)
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/builder/nested_eq.hpp:121]: probably, we just need to pass the custom comperator

File: include/sparrow/builder/builder_utils.hpp [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    concept is_nullable_like_generic = 
+    requires(T t)
+    {
+        //typename T::value_type;
+        { t.has_value() } -> std::convertible_to<bool>;
+        { t.get() };// -> std::convertible_to<typename T::value_type>;
+    };
+
+    template<class T>
+    concept is_nullable_like =(is_nullable_like_generic<T>  );//||  sparrow::is_nullable_v<T>);
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply include/sparrow/builder/builder_utils.hpp:105]: I was thinking of also adding std::optional support, but lets  this maybe later if at all

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**